src/main/java/appeng/client/gui/implementations/SpatialAnchorScreen.java

```diff
@@ -0,0 +1,257 @@
+import appeng.client.gui.widgets.Scrollbar;
+import appeng.client.gui.widgets.ServerSettingToggleButton;
+import appeng.client.gui.widgets.SettingToggleButton;
+import appeng.client.me.ItemRepo;
+import appeng.container.implementations.SpatialAnchorContainer;
+import appeng.core.localization.GuiText;
+import appeng.util.Platform;
+
+public class SpatialAnchorScreen extends AEBaseScreen<SpatialAnchorContainer> implements ISortSource {
+
+    private static final int SCROLL_AREA_WIDTH = 160;
+    private static final int SCROLL_AREA_HEIGHT = 78;
+    private static final int SCROLL_AREA_OFFSET_X = 0;
+    private static final int SCROLL_AREA_OFFSET_Y = 0;
+
+    private static final int SCROLL_COLUMNS = 3;
+    private static final int SCROLL_ROWS = 4;
+
+    private static final int SCROLL_ITEM_WIDTH = SCROLL_AREA_WIDTH / SCROLL_COLUMNS;
+    private static final int SCROLL_ITEM_HEIGHT = SCROLL_AREA_HEIGHT / SCROLL_ROWS;
+    private static final int SCROLL_ITEM_PADDING_X = 10;
+    private static final int SCROLL_ITEM_PADDING_Y = 3;
+
+    private final ItemRepo repo;
+    private final int rows = 4;
+
+    private int tooltip = -1;
+    private SettingToggleButton<YesNo> overlayToggle;
+
+    public SpatialAnchorScreen(SpatialAnchorContainer container, PlayerInventory playerInventory,
+            ITextComponent title) {
+        super(container, playerInventory, title);
+        this.ySize = 162;
+        this.xSize = 195;
+
+        final Scrollbar scrollbar = new Scrollbar();
+        this.setScrollBar(scrollbar);
+
+        this.repo = new ItemRepo(scrollbar, this);
+        this.repo.setRowSize(5);
+    }
+
+    @Override
+    public void init() {
+        super.init();
+        this.addButton(CommonButtons.togglePowerUnit(this.guiLeft - 18, this.guiTop + 8));
+        this.addButton(this.overlayToggle = new ServerSettingToggleButton<>(this.guiLeft - 18, this.guiTop + 28,
+                Settings.OVERLAY_MODE, YesNo.NO));
+    }
+
+    @Override
+    public void render(MatrixStack matrixStack, final int mouseX, final int mouseY, final float btn) {
+
+        final int gx = (this.width - this.xSize) / 2;
+        final int gy = (this.height - this.ySize) / 2;
+
+        this.tooltip = -1;
+
+        int y = 0;
+        int x = 0;
+        for (int z = 0; z <= SCROLL_ROWS * SCROLL_COLUMNS; z++) {
+            final int minX = gx + 9 + x * SCROLL_ITEM_WIDTH;
+            final int minY = gy + 40 + y * SCROLL_ITEM_HEIGHT;
+
+            if (minX + SCROLL_ITEM_PADDING_X < mouseX && minX + SCROLL_ITEM_WIDTH - SCROLL_ITEM_PADDING_X > mouseX) {
+                if (minY + SCROLL_ITEM_PADDING_Y < mouseY
+                        && minY + SCROLL_ITEM_HEIGHT - SCROLL_ITEM_PADDING_Y > mouseY) {
+                    this.tooltip = z;
+                    break;
+                }
+            }
+
+            x++;
+
+            if (x >= SCROLL_COLUMNS) {
+                y++;
+                x = 0;
+            }
+        }
+
```

Comment:
Minecraft UIs are such an ugly mess. Blaaaaaaaaaargh

Reply:
Yeah. And all these magic values, which are somehow related to eachother. Or not.